File: pipeline/src/storage/s3_client.py
"""Shared S3 client for the pipeline workers.

boto3 auto-reads AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY,
AWS_DEFAULT_REGION, and AWS_ENDPOINT_URL from the environment — no
explicit configuration code is needed here beyond reading the bucket
name.

Application code should go through the higher-level helpers in this
module (put_object, get_object, ping_s3) rather than constructing
boto3 commands directly, so tenant-isolation and error-logging
conventions stay in one place.

See docs/DECISIONS.md D002 and docs/STORAGE_LAYOUT.md.
"""
from __future__ import annotations

import logging
import os
import shutil
from pathlib import Path
from typing import Any, Optional

logger = logging.getLogger(__name__)

# Railway's R2/bucket service injects the bucket name as AWS_S3_BUCKET; older config used
# AWS_S3_BUCKET_NAME. Read AWS_S3_BUCKET first, fall back to the legacy name.
BUCKET = os.environ.get("AWS_S3_BUCKET") or os.environ.get("AWS_S3_BUCKET_NAME", "rfp-pipeline-local")

_s3_client: Optional[Any] = None

# ── Local filesystem driver (dev / sandbox) ─────────────────────────────────
#
# The FRONTEND has had one of these since STORAGE-LOCAL (lib/storage/s3-client.ts): with
# STORAGE_DRIVER=local the same helpers read and write under LOCAL_STORAGE_DIR instead of talking
# to R2. The pipeline never got the counterpart, so every storage call here needed boto3 — which is
# not installed in this sandbox. The failure was quiet rather than loud: the shredder catches the
# exception and logs "S3 fetch failed … skipping", so a document simply never gets extracted and
# the run reports success having done less.
#
# Same two variables, same layout (<dir>/<bucket>/<key>), so the two services agree about where a
# key lives and a file written by one is readable by the other. Production is untouched: without
# STORAGE_DRIVER=local every function takes the boto3 path exactly as before.
LOCAL = os.environ.get("STORAGE_DRIVER") == "local"
LOCAL_DIR = os.environ.get("LOCAL_STORAGE_DIR", "/tmp/govwin-storage")


def _local_path(key: str) -> Path:
    """Resolve a key to its on-disk path, refusing anything that escapes the bucket root.

    Containment is checked with is_relative_to, NOT a string prefix: with a bucket named
    `testbucket`, the key `../testbucket-evil/x` resolves outside the root yet still starts with
    it, so a prefix test would wave through the one thing this guard exists to stop.
    """
    root = (Path(LOCAL_DIR) / BUCKET).resolve()
    target = (root / key).resolve()
    if not target.is_relative_to(root):
        raise RuntimeError(f"storage key escapes the bucket root: {key}")
    return target


def get_s3_client() -> Any:
    """Lazy boto3 S3 client singleton.

    boto3 is imported lazily so that modules importing from
    ``src.storage`` can still be collected by pytest even when boto3
    is not installed in the test environment.
    """
    global _s3_client
    if _s3_client is None:
        import boto3  # type: ignore[import-not-found]

        _s3_client = boto3.client("s3")
    return _s3_client


def put_object(
    *,
    key: str,
    body: bytes,
    content_type: Optional[str] = None,
    cache_control: Optional[str] = None,
    metadata: Optional[dict[str, str]] = None,
) -> None:
    extra: dict[str, Any] = {}
    if content_type is not None:
        extra["ContentType"] = content_type
    if cache_control is not None:
        extra["CacheControl"] = cache_control
    if metadata is not None:
        extra["Metadata"] = metadata
    if LOCAL:
        # Resolved OUTSIDE the try: a refused key is a caller error with a specific message, and
        # collapsing it into the generic "storage put failed" would hide which of the two it was.
        path = _local_path(key)
        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_bytes(body)
            return
        except Exception as e:
            logger.error("[storage.put_object] local write failed key=%s err=%s", key, e)
            raise RuntimeError("storage put failed") from e
    try:
        get_s3_client().put_object(Bucket=BUCKET, Key=key, Body=body, **extra)
    except Exception as e:
        logger.error("[s3.put_object] failed key=%s err=%s", key, e)
        raise RuntimeError("storage put failed") from e


def get_object_bytes(key: str) -> Optional[bytes]:
    if LOCAL:
        path = _local_path(key)
        # A missing key is None on both drivers — the callers branch on that, not on an exception.
        return path.read_bytes() if path.is_file() else None
    try:
        res = get_s3_client().get_object(Bucket=BUCKET, Key=key)
    except Exception as e:
        name = getattr(e, "__class__", type(e)).__name__
        # boto3 raises ClientError with error codes; NoSuchKey is 404
        err = getattr(e, "response", {}).get("Error", {}) if hasattr(e, "response") else {}
        code = err.get("Code") if isinstance(err, dict) else None
        if name == "NoSuchKey" or code in ("NoSuchKey", "404", "NotFound"):
            return None
        logger.error("[s3.get_object_bytes] failed key=%s err=%s", key, e)
        raise RuntimeError("storage get failed") from e
    body = res.get("Body")
    if body is None:
        return None
    return body.read()


def object_exists(key: str) -> bool:
    if LOCAL:
        return _local_path(key).is_file()
    try:
        get_s3_client().head_object(Bucket=BUCKET, Key=key)
        return True
    except Exception as e:
        err = getattr(e, "response", {}).get("Error", {}) if hasattr(e, "response") else {}
        code = err.get("Code") if isinstance(err, dict) else None
        if code in ("NoSuchKey", "404", "NotFound"):
            return False
        logger.error("[s3.object_exists] failed key=%s err=%s", key, e)
        raise RuntimeError("storage head failed") from e


def delete_object(key: str) -> None:
    if LOCAL:
        # Deleting an absent key is a no-op on S3; keep that contract.
        _local_path(key).unlink(missing_ok=True)
        return
    try:
        get_s3_client().delete_object(Bucket=BUCKET, Key=key)
    except Exception as e:
        logger.error("[s3.delete_object] failed key=%s err=%s", key, e)
        raise RuntimeError("storage delete failed") from e


def ping_s3() -> dict[str, Any]:
    """Health check — verifies the bucket is reachable via HeadBucket."""
    if LOCAL:
        root = Path(LOCAL_DIR) / BUCKET
        try:
            root.mkdir(parents=True, exist_ok=True)
            return {"ok": True, "bucket": BUCKET, "driver": "local", "dir": str(root)}
        except Exception as e:
            return {"ok": False, "driver": "local", "error": str(e)}
    try:
        get_s3_client().head_bucket(Bucket=BUCKET)
        return {"ok": True, "bucket": BUCKET}
    except Exception as e:
        return {"ok": False, "error": str(e)}


# ── Convenience helpers (text, JSON, copy) ──────────────────────────


def put_text(*, key: str, text: str, metadata: Optional[dict[str, str]] = None) -> None:
    """Write a UTF-8 text file (markdown, extracted text, etc.)."""
    put_object(
        key=key,
        body=text.encode("utf-8"),
        content_type="text/markdown; charset=utf-8",
        metadata=metadata,
    )


def put_json(*, key: str, obj: Any, metadata: Optional[dict[str, str]] = None) -> None:
    """Write a JSON file (metadata, compliance snapshot, etc.)."""
    import json as _json

    put_object(
        key=key,
        body=_json.dumps(obj, indent=2, default=str).encode("utf-8"),
        content_type="application/json; charset=utf-8",
        metadata=metadata,
    )


def copy_object(*, source_key: str, dest_key: str) -> None:
    """Server-side copy within the same bucket (no download/upload)."""
    if LOCAL:
        src, dst = _local_path(source_key), _local_path(dest_key)  # refusals surface as themselves
        try:
            if not src.is_file():
                raise FileNotFoundError(source_key)
            dst.parent.mkdir(parents=True, exist_ok=True)
            shutil.copyfile(src, dst)
            return
        except Exception as e:
            logger.error("[storage.copy_object] local copy failed src=%s dst=%s err=%s",
                         source_key, dest_key, e)
            raise RuntimeError("storage copy failed") from e
    try:
        get_s3_client().copy_object(
            Bucket=BUCKET,
            CopySource={"Bucket": BUCKET, "Key": source_key},
            Key=dest_key,
        )
    except Exception as e:
        logger.error(
            "[s3.copy_object] failed src=%s dst=%s err=%s",
            source_key, dest_key, e,
        )
        raise RuntimeError("storage copy failed") from e


def list_keys(*, prefix: str, max_keys: int = 1000) -> list[str]:
    """List object keys under a prefix (for copy-all operations)."""
    if LOCAL:
        root = (Path(LOCAL_DIR) / BUCKET).resolve()
        if not root.is_dir():
            return []
        # Keys are POSIX-style and relative to the bucket root, matching what S3 returns.
        keys = sorted(
            str(p.relative_to(root)).replace(os.sep, "/")
            for p in root.rglob("*") if p.is_file()
        )
        return [k for k in keys if k.startswith(prefix)][:max_keys]
    try:
        resp = get_s3_client().list_objects_v2(
            Bucket=BUCKET, Prefix=prefix, MaxKeys=max_keys,
        )
        return [obj["Key"] for obj in resp.get("Contents", [])]
    except Exception as e:
        logger.error("[s3.list_keys] failed prefix=%s err=%s", prefix, e)
        raise RuntimeError("storage list failed") from e

File: frontend/lib/storage/s3-client.ts
/**
 * Shared S3 client for the frontend (Next.js server routes + server
 * components). The AWS SDK auto-reads AWS_ACCESS_KEY_ID,
 * AWS_SECRET_ACCESS_KEY, AWS_DEFAULT_REGION, and AWS_ENDPOINT_URL from
 * the environment — no explicit configuration code is needed here.
 *
 * Application code should go through the higher-level helpers in
 * this module (putObject, getObject, getSignedGetUrl) rather than
 * constructing command objects directly, so tenant-isolation and
 * error-logging conventions stay in one place.
 *
 * LOCAL DRIVER (dev/sandbox): with STORAGE_DRIVER=local, these SAME helpers are
 * backed by the local filesystem instead of a reachable S3/R2 bucket — so the
 * image pipeline (capture, box-atomize, uploads, export) runs end-to-end where
 * Railway's R2 service isn't linked. Keys are stored verbatim (S3 naming) under
 * LOCAL_STORAGE_DIR/<bucket>/; signed URLs resolve to the gated
 * /api/storage/local/<key> route. Production (no flag) is byte-for-byte unchanged.
 *
 * See docs/DECISIONS.md D002 and docs/STORAGE_LAYOUT.md.
 */
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
  HeadObjectCommand,
  HeadBucketCommand,
  DeleteObjectCommand,
  CopyObjectCommand,
  ListObjectsV2Command,
  type PutObjectCommandInput,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { promises as fs } from 'fs';
import { join, dirname } from 'path';

// ── Local driver flag + primitives ──────────────────────────────────────────
export const LOCAL = process.env.STORAGE_DRIVER === 'local';
const LOCAL_DIR = process.env.LOCAL_STORAGE_DIR || '/tmp/govwin-storage';
const LOCAL_URL_BASE = '/api/storage/local';

// Railway's R2/bucket service injects the bucket name as AWS_S3_BUCKET; older config used
// AWS_S3_BUCKET_NAME. Read AWS_S3_BUCKET first, fall back to the legacy name — otherwise the
// code looks for a var Railway never sets and throws below (the recurring prod storage error).
const _bucketEnv = process.env.AWS_S3_BUCKET || process.env.AWS_S3_BUCKET_NAME;

// Skip the guard during the Next.js "Collecting page data" step at
// build time — Railway's build container has no runtime secrets.
// Runtime still throws if the var is missing when the request fires (unless local).
const _isBuildPhase = process.env.NEXT_PHASE === 'phase-production-build';
if (!_bucketEnv && process.env.NODE_ENV === 'production' && !_isBuildPhase && !LOCAL) {
  throw new Error('[storage/s3-client] AWS_S3_BUCKET (or AWS_S3_BUCKET_NAME) is required in production');
}

export const BUCKET = _bucketEnv || 'rfp-pipeline-local';

const _localPath = (key: string) => join(LOCAL_DIR, BUCKET, key);
async function _localWrite(key: string, body: Buffer | Uint8Array | string, contentType?: string): Promise<void> {
  const p = _localPath(key);
  await fs.mkdir(dirname(p), { recursive: true });
  const buf = Buffer.isBuffer(body) ? body : typeof body === 'string' ? Buffer.from(body) : Buffer.from(body);
  await fs.writeFile(p, buf);
  await fs.writeFile(`${p}.contenttype`, contentType || 'application/octet-stream');
}
async function _localRead(key: string): Promise<Buffer | null> {
  try { return await fs.readFile(_localPath(key)); } catch { return null; }
}
/** Content-type sidecar for a locally-stored key (used by the serving route). */
export async function localContentType(key: string): Promise<string> {
  try { return (await fs.readFile(`${_localPath(key)}.contenttype`, 'utf8')).trim() || 'application/octet-stream'; }
  catch { return 'application/octet-stream'; }
}
/** Read/write helpers the local serving route uses (never touched in prod). */
export async function localReadObject(key: string): Promise<Buffer | null> { return _localRead(key); }
export async function localWriteObject(key: string, body: Buffer, contentType?: string): Promise<void> { return _localWrite(key, body, contentType); }

// Singleton — construct once per process. The SDK v3 reads credentials
// from AWS_ACCESS_KEY_ID + AWS_SECRET_ACCESS_KEY and endpoint from
// AWS_ENDPOINT_URL automatically. HOWEVER: it reads region from
// AWS_REGION, not AWS_DEFAULT_REGION (which is what Railway + boto3
// use). We bridge that here. (Unused when LOCAL — construction is lazy/harmless.)
export const s3 = new S3Client({
  forcePathStyle: true,
  region: process.env.AWS_DEFAULT_REGION || process.env.AWS_REGION || 'auto',
});

export interface PutObjectInput {
  key: string;
  body: Buffer | Uint8Array | string;
  contentType?: string;
  cacheControl?: string;
  metadata?: Record<string, string>;
}

export async function putObject(input: PutObjectInput): Promise<void> {
  if (LOCAL) return _localWrite(input.key, input.body, input.contentType);
  const params: PutObjectCommandInput = {
    Bucket: BUCKET,
    Key: input.key,
    Body: input.body,
    ContentType: input.contentType,
    CacheControl: input.cacheControl,
    Metadata: input.metadata,
  };
  try {
    await s3.send(new PutObjectCommand(params));
  } catch (e) {
    const detail = e instanceof Error ? e.message : String(e);
    const code = (e as { Code?: string; name?: string })?.Code ?? (e as { name?: string })?.name ?? 'unknown';
    console.error('[s3.putObject] failed', { key: input.key, bucket: BUCKET, code, err: detail });
    throw new Error(`S3 put failed (${code}): ${detail}`);
  }
}

export async function getObjectBuffer(key: string): Promise<Buffer | null> {
  if (LOCAL) return _localRead(key);
  try {
    const res = await s3.send(new GetObjectCommand({ Bucket: BUCKET, Key: key }));
    if (!res.Body) return null;
    const chunks: Buffer[] = [];
    // @ts-expect-error Node stream — SDK's Body is a Readable in Node runtime
    for await (const chunk of res.Body) {
      chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
    }
    return Buffer.concat(chunks);
  } catch (e) {
    const name = (e as { name?: string } | null)?.name;
    if (name === 'NoSuchKey' || name === 'NotFound') return null;
    console.error('[s3.getObjectBuffer] failed', { key, err: String(e) });
    throw new Error('storage get failed');
  }
}

export async function objectExists(key: string): Promise<boolean> {
  if (LOCAL) { try { await fs.access(_localPath(key)); return true; } catch { return false; } }
  try {
    await s3.send(new HeadObjectCommand({ Bucket: BUCKET, Key: key }));
    return true;
  } catch (e) {
    const name = (e as { name?: string } | null)?.name;
    if (name === 'NotFound' || name === 'NoSuchKey') return false;
    console.error('[s3.objectExists] failed', { key, err: String(e) });
    throw new Error('storage head failed');
  }
}

export async function deleteObject(key: string): Promise<void> {
  if (LOCAL) { try { await fs.unlink(_localPath(key)); await fs.unlink(`${_localPath(key)}.contenttype`); } catch { /* absent is fine */ } return; }
  try {
    await s3.send(new DeleteObjectCommand({ Bucket: BUCKET, Key: key }));
  } catch (e) {
    console.error('[s3.deleteObject] failed', { key, err: String(e) });
    throw new Error('storage delete failed');
  }
}

/**
 * Copy an object within the same bucket.
 * Used during proposal provisioning to copy RFP documents into the customer sandbox.
 */
export async function copyObject({ sourceKey, destKey }: { sourceKey: string; destKey: string }): Promise<void> {
  if (LOCAL) {
    const b = await _localRead(sourceKey);
    if (b == null) throw new Error(`S3 copy failed (local): source missing ${sourceKey}`);
    await _localWrite(destKey, b, await localContentType(sourceKey));
    return;
  }
  try {
    const command = new CopyObjectCommand({
      Bucket: BUCKET,
      CopySource: `${BUCKET}/${sourceKey}`,
      Key: destKey,
    });
    await s3.send(command);
  } catch (e) {
    const detail = e instanceof Error ? e.message : String(e);
    const code = (e as { Code?: string; name?: string })?.Code ?? (e as { name?: string })?.name ?? 'unknown';
    console.error('[s3.copyObject] failed', { sourceKey, destKey, bucket: BUCKET, code, err: detail });
    throw new Error(`S3 copy failed (${code}): ${detail}`);
  }
}

/**
 * Presigned GET URL for time-limited browser downloads.
 * Default TTL is 15 minutes. (LOCAL: a stable gated route URL — the local analog of a bearer link.)
 */
export async function getSignedGetUrl(key: string, expiresInSeconds = 900): Promise<string> {
  if (LOCAL) return `${LOCAL_URL_BASE}/${key.split('/').map(encodeURIComponent).join('/')}`;
  try {
    return await getSignedUrl(
      s3,
      new GetObjectCommand({ Bucket: BUCKET, Key: key }),
      { expiresIn: expiresInSeconds },
    );
  } catch (e) {
    console.error('[s3.getSignedGetUrl] failed', { key, err: String(e) });
    throw new Error('storage sign failed');
  }
}

/**
 * Presigned PUT URL for direct browser-to-S3 uploads.
 * Bypasses Next.js body limits — critical for large files (300MB+).
 * Default TTL is 15 minutes. (LOCAL: the gated route accepts the PUT to the filesystem.)
 */
export async function getSignedPutUrl(
  key: string,
  contentType?: string,
  expiresInSeconds = 900,
): Promise<string> {
  if (LOCAL) return `${LOCAL_URL_BASE}/${key.split('/').map(encodeURIComponent).join('/')}`;
  try {
    return await getSignedUrl(
      s3,
      new PutObjectCommand({
        Bucket: BUCKET,
        Key: key,
        ContentType: contentType || 'application/octet-stream',
      }),
      { expiresIn: expiresInSeconds },
    );
  } catch (e) {
    console.error('[s3.getSignedPutUrl] failed', { key, err: String(e) });
    throw new Error('storage presign-put failed');
  }
}

/**
 * Health check — verifies the bucket is reachable.
 * Used by /api/health and pipeline /healthz.
 */
export async function pingS3(): Promise<{ ok: true; bucket: string } | { ok: false; error: string }> {
  if (LOCAL) return { ok: true, bucket: `${BUCKET} (local fs)` };
  try {
    await s3.send(new HeadBucketCommand({ Bucket: BUCKET }));
    return { ok: true, bucket: BUCKET };
  } catch (e) {
    return { ok: false, error: String(e) };
  }
}

/**
 * List objects and common prefixes under a given S3 prefix.
 * Used by the admin file manager to browse the bucket.
 */
export async function listObjects(prefix: string, delimiter = '/'): Promise<{
  objects: Array<{ key: string; size: number; lastModified: Date | null }>;
  prefixes: string[];
}> {
  if (LOCAL) {
    // Walk LOCAL_DIR/<bucket>/<prefix>; one level deep per the delimiter convention.
    const base = join(LOCAL_DIR, BUCKET);
    const dir = join(base, prefix);
    const objects: Array<{ key: string; size: number; lastModified: Date | null }> = [];
    const prefixes: string[] = [];
    try {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const e of entries) {
        if (e.name.endsWith('.contenttype')) continue;
        const rel = `${prefix}${prefix && !prefix.endsWith('/') ? '/' : ''}${e.name}`;
        if (e.isDirectory()) { if (delimiter) prefixes.push(`${rel}/`); }
        else { const st = await fs.stat(join(dir, e.name)); objects.push({ key: rel, size: st.size, lastModified: st.mtime }); }
      }
    } catch { /* missing prefix → empty */ }
    return { objects, prefixes };
  }
  if (!s3 || !BUCKET) return { objects: [], prefixes: [] };
  const cmd = new ListObjectsV2Command({
    Bucket: BUCKET,
    Prefix: prefix,
    Delimiter: delimiter,
  });
  let res;
  try {
    res = await s3.send(cmd);
  } catch (e) {
    const detail = e instanceof Error ? e.message : String(e);
    console.error('[s3.listObjects] failed', { prefix, bucket: BUCKET, err: detail });
    throw new Error(`S3 list failed: ${detail}`);
  }
  const objects = (res.Contents ?? [])
    .filter(o => o.Key && o.Key !== prefix)
    .map(o => ({
      key: o.Key!,
      size: o.Size ?? 0,
      lastModified: o.LastModified ?? null,
    }));
  const prefixes = (res.CommonPrefixes ?? [])
    .map(p => p.Prefix!)
    .filter(Boolean);
  return { objects, prefixes };
}
